### ISSUE TITLE
no-inner-html rule: false positive for non-jQuery scenario

### ISSUE THREAD

**Issue by amacleay** [Microsoft/tslint-microsoft-contrib#425]
I have a project using:
* enzyme
* chai
* chai-enzyme
* mocha-webpack

I have something similar to the following in a test:
```javascript
// $ cat test/AuthenticatedApp.test.tsx
import React from 'react';                                                                                                                                               
import { expect } from 'chai';                                                                                                                                           
import { render } from 'enzyme';
import AuthenticatedApp from '../src/AuthenticatedApp.tsx';

describe('AuthenticatedApp', function() {
  context('first', function() {
    it('renders children', function() {
      const app = render(
        <AuthenticatedApp session={{}}>
          <h1>Hi guys</h1>
        </AuthenticatedApp>,
      );
      expect(app).to.have.html('<h1>Hi guys</h1>');  // ERROR
    });
  });
});
```

I get the following error:
```
test/AuthenticatedApp.test.tsx14, 7]: Using the html() function to write a string to innerHTML is insecure: expect(app).to.have.html('<h1>Hi guys</h1>')
```

This is incorrect behavior. I am not using the [.html](https://api.jquery.com/html/) function, I am using a [chai-enzyme assertion](https://github.com/producthunt/chai-enzyme#htmlstr)

I believe this could be fixed by adding an `isJquery` check alongside https://github.com/Microsoft/tslint-microsoft-contrib/blob/master/src/noInnerHtmlRule.ts#L57

Workaround: I have disabled `no-inner-html` in my `test` directory.

**Comment by JoshuaKGoldberg** [comment]
Hey, thanks for filing this - agreed that we need a way to exclude test files. However, hardcoding it to `jquery|$` isn't the right way to go for all projects. What about projects that use other DOM libraries, or have jQuery aliased under different names?

Instead of hardcoding it to `jquery|$`, I think we should allow the user to provide a matcher (stringified RegExp, presumably) that filters what gets warned on. Your use case would provide something like `jquery|$`.

Thoughts?

**Comment by amacleay** [comment]
Seems probable, but not totally certain.

This part of the rule is trying to enforce that no method gets called which can directly replace the HTML of an element. It so happens that jQuery instances have a method named `html` that does that, but another DOM processing library might have different methods that do the same thing. Worse, another DOM processing library might expose that facility through a simple function interface (`replaceHTML(selector, contents)`). The general case of this problem is beyond the scope of one rule.

To avoid having this rule try to handle any general case, I had considered that potentially the rule should be split in two: one rule which forbids assignment to `{inner,outer}HTML` (no user-defined parameters) and another rule that forbids calls with arguments to the `html` method on a jquery-looking expression (which could take a user-defined matcher).

This second rule would still have problems: the first and most obvious of which is that string matching for the method invocant is error-prone and at best an approximation. Maybe a bigger issue: why limit it to just methods called `html`? There might be a library that has a method called `setHtml`. And then, if we're considering that the method name might be parametrizable, maybe this rule is actually just a general rule that takes a list of forbidden invocant/method pairs and ensures that these methods are never called, or are always called with a certain number of arguments (so the jQuery.html case would be the special case of this rule where the parameters look like `{"invocant": "jquery|jQuery|$", "method": "html", "argsLength": "1+"}`).

I just mean to point out that there are a number of different ways to go with this, and it's a matter of striking the right balance of ease-of-use, reliability, and complexity.

What do you think?
